File: core/cpp/test/default/common/topological_sort_test.cc
// Copyright (c) 2020 The DashQL Authors

#include <sstream>

#include "dashql/common/topological_sort.h"
#include "gtest/gtest.h"

using namespace std;
using namespace dashql;

namespace {

enum class TestOpType {
    POP,
    DEC
};

struct TestOp {
    TestOpType type;
    unsigned value;
};

std::pair<unsigned, unsigned> PUSH(unsigned key, unsigned rank) { return { key, rank }; }
TestOp POP(unsigned value) { return { TestOpType::POP, value }; }
TestOp DEC(unsigned value) { return { TestOpType::DEC, value }; }

struct TopoSortTest {
    std::vector<std::pair<unsigned, unsigned>> input;
    std::vector<TestOp> ops;
};

struct TopologicalSortTestSuite: ::testing::TestWithParam<TopoSortTest> {};

TEST_P(TopologicalSortTestSuite, SequenceMatches) {
    auto& param = GetParam();
    TopologicalSort<unsigned> heap{param.input};
    for (auto [type, value]: param.ops) {
        if (type == TestOpType::DEC) {
            heap.DecrementKey(value);
        } else {
            ASSERT_FALSE(heap.Empty());
            ASSERT_EQ(std::get<0>(heap.Top()), value);
            heap.Pop();
        }
    }
    ASSERT_TRUE(heap.Empty());
}

INSTANTIATE_TEST_SUITE_P(ToplogicalSort, TopologicalSortTestSuite, ::testing::Values(
    TopoSortTest{},
    TopoSortTest{
        {PUSH(0, 0)},
        {POP(0)}
    },
    TopoSortTest{
        {PUSH(0, 2), PUSH(1,1)},
        {POP(1), POP(0)}
    },
    TopoSortTest{
        {PUSH(0, 0), PUSH(1,2), PUSH(2, 1), PUSH(3, 1)},
        {POP(0), DEC(1), DEC(2), POP(2), POP(1), POP(3)}
    }
));

}  // namespace

File: core/test/utils/topological_sort.test.ts
import * as core from '../../src/index_node';

enum TestOpType { POP, DEC }
interface TestOp {
    type: TestOpType;
    key: number;
    by: number;
}

function PUSH(key: number, rank: number): [number, number] { return [key, rank]; }
function POP(key: number) { return { type: TestOpType.POP, key: key, by: 0 }; }
function DEC(key: number, by: number = 1) { return { type: TestOpType.DEC, key: key, by: by }; }

const tests: [string, [number, number][], TestOp[]][] = [
    [
        "simple_1",
        [ PUSH(0, 0) ],
        [ POP(0) ]
    ],
    [ 
        "simple_2",
        [ PUSH(0, 2), PUSH(1, 1) ],
        [ POP(1), POP(0) ]
    ],
    [ 
        "simple_3",
        [ PUSH(0, 0), PUSH(1,2), PUSH(2, 1), PUSH(3, 1) ],
        [ POP(0), DEC(1), DEC(2), POP(2), POP(1), POP(3) ]
    ],
];

describe('TopologicalSort', () => {
    test.each(tests)(
        "%s",
        (_name, input, ops) => {
            input.sort((l, r) => l[1] - r[1]);
            const heap = new core.utils.TopologicalSort(input);
            for (const op of ops) {
                if (op.type == TestOpType.DEC) {
                    heap.decrementKey(op.key, op.by);
                } else {
                    expect(heap.empty()).toBe(false);
                    expect(heap.top()).toBe(op.key);
                    heap.pop();
                }
            }
        }
    );
})
